### ISSUE TITLE
WSL bind mounts being treated as volume mounts in some instances

### ISSUE THREAD

**Issue by apm963** [docker/for-win#12654]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 07E11884-7E06-4A44-81A4-840C1CD420A3/20220408154736

### Actual behavior

Depending on how the `docker.sock` is made available, a container created with a bind mount will either be a host bind mount or a volume mount. Creating a container via TCP (exposed daemon on port 2375) will always result in incorrect bind mount behavior. This does not affect Hyper-V or Docker for Mac.

### Expected behavior

Bind mounts will always be bind mounts and will not be volume mounts. Bind mounts will mount directories on the host (WSL) filesystem.

### Information

This behavior has been reproduced on two separate Windows 11 Home machines. A similar (if not the same) behavior was experienced at least a year ago on a Windows 10 Home machine.

  - Windows Version: Windows 11 Home 21H2
  - Docker Desktop Version: 4.6.1 (76265)
  - WSL2 or Hyper-V backend?  WSL2
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

After much investigation, it has been determined that using a bind-mounted `docker.sock` has a different behavior depending on the way the container it is being used in was created.

* Container created via `docker run`: Running `mount | grep docker.sock` reveals the mount point is using device `none` with `tmpfs` driver. This method allows normal, unimpeded container bind-mount creation.
* Container created via a Docker Swarm deployment: Running `mount | grep docker.sock` shows the mount point is using device `overlay` with `overlay` driver. This method causes undesirable behavior when executing `docker run -v /a:/b ...` bind-mounts.

The issue is that when you execute `docker` commands against a `docker.sock` that is mounted using the `overlay` driver, the behavior is different. Specifically notable is creating a container with a host bind-mount. When you `docker run` (including using `docker-compose up` and similar) with the `-v /host:/container` flag, the newly created container's mount does not map to the *host*'s filesystem (in this case, the default WSL distro). Instead, the mount point uses an `overlay` driver and is treated as a *volume* mount, not a bind mount (which would use a different mount type \[eg. `ext4`\] and device \[eg. `/dev/sdd`\]).

This behavior is not revealed by comparing the two containers created above using `docker inspect`. From what I could find, this is not documented behavior. This behavior is only experienced on WSL and not on native linux hosts, nor Hyper-V or Docker for Mac.

In summary, containers that are created via Swarm (or daemon on exposed TCP) and have `docker.sock` bind mounted within them result in different `docker run -v` behavior than containers created manually (non-Swarm).

Also see [this Stack Exchange issue](https://superuser.com/questions/1576832/docker-desktop-wsl2-volume-mapping-not-working) which sounds like the same problem.

### Output of `& "C:\Program Files\Docker\Docker\resources\com.docker.diagnose.exe" check`

```text
[2022-04-08T15:49:00.792379900Z][com.docker.diagnose.exe][I] ipc.NewClient: 3b8ccc4e-com.docker.diagnose -> \\.\pipe\dockerBackendV2 com.docker.service.exe
[goroutine 1 [running, locked to thread]:
[win/backend/pkg/service.NewClientForPath(...)
[       win/backend/pkg/service/service.go:42
[win/backend/pkg/service.NewClient({0xa414c4, 0x13}, {0x0, 0x0, 0x0})
[       win/backend/pkg/service/service.go:31 +0xd0
[common/pkg/diagkit/gather/diagnose.init()
[       common/pkg/diagkit/gather/diagnose/ipc_windows.go:23 +0x11f
Starting diagnostics

[PASS] DD0027: is there available disk space on the host?
[PASS] DD0028: is there available VM disk space?
[PASS] DD0031: does the Docker API work?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0001: is the application running?
[SKIP] DD0018: does the host support virtualization?
[PASS] DD0002: does the bootloader have virtualization enabled?
[PASS] DD0017: can a VM be started?
[PASS] DD0024: is WSL installed?
[PASS] DD0021: is the WSL 2 Windows Feature enabled?
[PASS] DD0022: is the Virtual Machine Platform Windows Feature enabled?
[PASS] DD0025: are WSL distros installed?
[PASS] DD0026: is the WSL LxssManager service running?
[PASS] DD0029: is the WSL 2 Linux filesystem corrupt?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0003: is the Docker CLI working?
[PASS] DD0013: is the $PATH ok?
[PASS] DD0005: is the user in the docker-users group?
[PASS] DD0007: is the backend responding?
[PASS] DD0014: are the backend processes running?
[PASS] DD0008: is the native API responding?
[PASS] DD0009: is the vpnkit API responding?
[PASS] DD0010: is the Docker API proxy responding?
[PASS] DD0006: is the Docker Desktop Service responding?
[PASS] DD0012: is the VM networking working?
[PASS] DD0032: do Docker networks overlap with host IPs?
[SKIP] DD0030: is the image access management authorized?
[PASS] DD0033: does the host have Internet access?
No fatal errors detected.
```

### Steps to reproduce the behavior

Prerequisites:

1. A host that is set up as a swarm manager
2. A Windows host that is set up as a swarm worker node. Docker Desktop must be in WSL mode.
3. The Windows host should have a WSL distro set up. Try Debian. Make sure this distro is set as the default (`wsl --setdefault DISTRONAME`) and configured to have WSL integration (in Docker Desktop WSL integration settings)

Deploy a stack on the swarm manager with the following config:

```yml
version: "3.4"

services:
  wsl_repro:
    image: alpine
    command: ["sleep", "6000"]
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /var/run/docker.sock:/tmp/docker.sock     # optional
      - /sys/fs/cgroup/:/sys/fs/cgroup/:ro        # optional
    deploy:
      restart_policy:
        condition: on-failure
        delay: 5s
        max_attempts: 1
        window: 120s
      mode: global
      placement:
        constraints:
          - node.role != manager
```

Steps to perform on the WSL host:

1. Wait for stack to be deployed and the alpine container to come up
2. `exec` into the alpine container. We will call this the "swarm container."
3. Run `mount | grep docker.sock`. Note this is using the `overlay` driver (should not be).
4. Open a new shell for the WSL distro (eg. debian shell)
5. In the WSL shell, run `docker run --rm -it -v /var/run/docker.sock:/var/run/docker.sock alpine /bin/sh`. We will call this the "manual container."
6. Run `mount | grep docker.sock`. Note this is not using the `overlay` driver but is instead using `tmpfs` with `none` device.
7. Going back to the swarm container shell, run `docker run --rm -it -v /home:/mnt/host:ro alpine /bin/sh -c '(mount | grep mnt/host) ; ls -l /mnt/host'`. Keep those results for comparison.
8. In the manual container shell, run the same command as in step 7.
9. Compare the output from the two commands above. You will note that the `mount` is `overlay` in the swarm container output but is `ext4` (or similar) in the manual container output. You will also likely see a difference in the `ls` output - the manual container shell will show the WSL distro's `home` dir whereas the swarm container shell's output will be empty.
10. Optional: Repeat steps 7 and 8 but modify the command to not remove the container on stop. Use `docker inspect` on both containers that were created and diff. You will notice the `Mounts` and other mount-related properties are identical. From what the Docker daemon tells us, these two containers are configured the same way and should have the same behavior. Based on step 9, they are in fact different.

As this shows, the swarm container's `docker run -v` bind-mount does not mount to the host but instead using the overlay driver like a volume mount would. If you were to create files in here, they would be stored in the volume mount WSL location.

> Note the `--mount` flag results in the same behavior. The `-v` flag is being used here for brevity.

It is interesting to note that this same `overlay` behavior occurs when interacting with docker over an exposed TCP port (there is a checkbox in the Docker Desktop settings for this). Try the same steps as above, but instead use `docker -H host.docker.internal:2376 run ...` for the steps that apply to the manually created container. You will note that all bind mounted locations are using the `overlay` driver.

In summary, if `docker.sock` is bind mounted but is `mount`ed in the resulting container with an `overlay` driver then any host bind mounts will be treated as volumes. It has not been determined how to directly bind mount the host filesystem when in this state. By contrast, a container that is created manually and has `docker.sock` properly bind mounted (not using the `overlay` driver) can interact with the docker daemon unimpeded.

**Comment by mateuszdrab** [comment]
Ahh, so this is why I spent the last half an hour trying to figure out why a bind mounted folder from my Windows host is empty despite the container writing to it.

It worked for me before though - suggesting it could be a regression.

I tested it now and I can only create a container from Ubuntu in WSL2 to have the folder bind mounted correctly. Attempts from Windows docker client or using the sock in any other way like through the Portainer Agent, result in the overlay.

Thank you for figuring it out
